src/dev/HallSensor.cpp

```diff
@@ -0,0 +1,93 @@
+#include <EVT/io/GPIO.hpp>
+#include <HALf3/stm32f3xx_hal.h>
+#include <dev/HallSensor.hpp>
+
+namespace IO = EVT::core::IO;
+
+// TODO Tune these values
+constexpr uint32_t THRESHOLD = 10000;// Threshold for wheel speed
+constexpr uint32_t THETA = 2;        // Constant for wheel speed calculation
+
+namespace WSS::DEV {
+
+HallSensor::HallSensor(IO::GPIO& gpio, uint32_t wheelRadius) : gpio((IO::GPIO&) gpio), wheelRadius(wheelRadius) {
+    this->prevTime = 0;
+    this->lastInterval = 0;
+    this->state = WheelSpeedState::STOP;
+    this->magnetInLastRead = false;
+}
+
+void HallSensor::update() {
+    uint32_t now = HAL_GetTick();
+
+    switch (state) {
+    case WheelSpeedState::STOP:
+        if (gpio.readPin() == MAGNET_DETECTED_STATE) {
+            // If the magnet is detected for the first time, record the time and start initializing
+            if (!magnetInLastRead) {
+                prevTime = now;
+                magnetInLastRead = true;
+                state = WheelSpeedState::INITIALIZING;
+            }
+        } else {
+            magnetInLastRead = false;
+        }
+        break;
+
+    case WheelSpeedState::INITIALIZING:
+        if (gpio.readPin() == MAGNET_DETECTED_STATE) {
+            if (!magnetInLastRead) {
+                // If the magnet is detected for the second time, record the time and the first
+                // interval and start maintaining
+                lastInterval = now - prevTime;
+                prevTime = now;
+                magnetInLastRead = true;
+                state = WheelSpeedState::MAINTAIN;
+            }
+        } else {
+            magnetInLastRead = false;
+        }
+
+        // If the magnet hasn't been detected for a significant amount of time, go back to the stop
+        // state
+        if (now - prevTime > THRESHOLD) {
+            state = WheelSpeedState::STOP;
+            prevTime = 0;
+        }
+        break;
+
+    case WheelSpeedState::MAINTAIN:
+        if (gpio.readPin() == MAGNET_DETECTED_STATE) {
+            if (!magnetInLastRead) {
+                // If the magnet is newly detected, record the time and interval
+                lastInterval = now - prevTime;
+                prevTime = now;
+                magnetInLastRead = true;
+            }
+        } else {
+            magnetInLastRead = false;
+        }
+
+        uint32_t elapsedTime = now - prevTime;
+        // If the magnet hasn't been detected for a significant amount of time, go back to the stop
+        // state
+        if (elapsedTime > THRESHOLD) {
```

Comment:
why is the elapsedTime math done not inline, whereas on line 53 of this file (a lil up from here), its done inline.